Mathlib/Tactic/Linter/HaveLetLinter.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean.Elab.Command
+import Lean.Server.InfoUtils
+
+/-!
+#  The `have` vs `let` linter
+
+The `have` vs `let` linter flags uses of `have` to introduce a hypothesis whose Type is not `Prop`.
+
+The option for this linter is a natural number, but really there are only 3 settings:
+* `0` -- inactive;
+* `1` -- active only on noisy declarations;
+* `2` or more -- always active.
+
+TODO:
+* Also lint `let` vs `have`.
+* `haveI` may need to change to `let/letI`?
+* `replace`, `classical!`, `classical`, `tauto` internally use `have`:
+  should the linter act on them as well?
+-/
+
+open Lean Elab Command Meta
+
+namespace Mathlib.Linter
+
+/-- The `have` vs `let` linter emits a warning on `have`s introducing a hypothesis whose
+Type is not `Prop`.
+There are three settings:
+* `0` -- inactive;
```

Comment:
Is this just a limitation of the types that can be passed to a built in option?

Reply:
Essentially yes. A `Lean.Option` is
```lean
/-- A strongly-typed reference to an option. -/
protected structure Option (α : Type) where
  name     : Name
  defValue : α
  deriving Inhabited
```
with an arbitrary type `α`.  However, the `Option` is then stored in a `KVMap`, whose `Value`s need to be in a
```lean
/-- Value stored in a key-value map. -/
inductive DataValue where
  | ofString (v : String)
  | ofBool   (v : Bool)
  | ofName   (v : Name)
  | ofNat    (v : Nat)
  | ofInt    (v : Int)
  | ofSyntax (v : Syntax)
  deriving Inhabited, BEq, Repr
```

Reply:
I guess it is discoverable but it would nice to have custom syntax for verbosity of things.